textattack/search_methods/PSO_algorithm.py

```diff
@@ -0,0 +1,266 @@
+        neighbors_list = [[] for _ in range(len(words))]
+        transformations = self.get_transformations(
+            attacked_text, original_text=self.original_attacked_text
+        )
+        for transformed_text in transformations:
+            try:
+                diff_idx = attacked_text.first_word_diff_index(transformed_text)
+                neighbors_list[diff_idx].append(transformed_text.words[diff_idx])
+            except:
+                assert len(attacked_text.words) == len(transformed_text.words)
+                assert all(
+                    [
+                        w1 == w2
+                        for w1, w2 in zip(attacked_text.words, transformed_text.words)
+                    ]
+                )
+        neighbors_list = [np.array(x) for x in neighbors_list]
+        return neighbors_list
+
+    def equal(self, a, b):
+        if a == b:
+            return -3
+        else:
+            return 3
+
+    def turn(self, x1, x2, prob, x_len):
+        indices_to_replace = []
+        words_to_replace = []
+        x2_words = x2.words
+        for i in range(x_len):
+            if np.random.uniform() < prob[i]:
+                indices_to_replace.append(i)
+                words_to_replace.append(x2_words[i])
+        new_text = x1.replace_words_at_indices(indices_to_replace, words_to_replace)
+        return new_text
+
+    def count_change_ratio(self, x1, x2, x_len):
+        change_ratio = float(np.sum(x1.words != x2.words)) / float(x_len)
+        return change_ratio
+
+    def sigmod(self, n):
+        return 1 / (1 + np.exp(-n))
+
+    def _perform_search(self, initial_result):
+        self.original_attacked_text = initial_result.attacked_text
+        x_len = len(self.original_attacked_text.words)
+        self.correct_output = initial_result.output
+
+        # get word substitute candidates and generate population
+        neighbors_list = self._get_neighbors_list(self.original_attacked_text)
+        neighbors_len = [len(x) for x in neighbors_list]
+        pop = self.generate_population(
+            self.original_attacked_text, neighbors_list, neighbors_len
+        )
+
+        # test population against target model
+        pop_results, self.search_over = self.get_goal_results(pop, self.correct_output)
+        if self.search_over:
+            return max(pop_results, key=lambda x: x.score)
+        pop_scores = np.array([r.score for r in pop_results])
+
+        # rank the scores from low to high and check if there is a successful attack
+        part_elites = deepcopy(pop)
+        part_elites_scores = pop_scores
+        top_attack = np.argmax(pop_scores)
+        all_elite = pop[top_attack]
+        all_elite_score = pop_scores[top_attack]
+        if pop_results[top_attack].succeeded:
+            return pop_results[top_attack]
+
+        # set up hyper-parameters
+        Omega_1 = 0.8
+        Omega_2 = 0.2
+        C1_origin = 0.8
+        C2_origin = 0.2
```

Comment:
I think these should all be properties of the class, initialized in `__init__`

```
self.Omega_1 = ...
```

because someone might want to try running this PSO algorithm with different hyperparameters!